frame/staking/src/pallet/mod.rs

```diff
@@ -559,6 +651,184 @@ pub mod pallet {
+			let claimed_pages = ClaimedRewards::<T>::get(era, validator);
+			let claimed_page_count = claimed_pages.len() as PageIndex;
+
+			// find the first page that is not claimed.
+			for page in 0..claimed_page_count as PageIndex {
+				debug_assert!(page <= claimed_pages[page as usize]);
+				if page < claimed_pages[page as usize] {
+					return Some(page)
+				}
+			}
+			// all pages are claimed
+			return if claimed_page_count < page_count { Some(claimed_page_count) } else { None }
+		}
+
+		/// Checks if exposure is paged or not.
+		fn is_non_paged_exposure(era: EraIndex, validator: &T::AccountId) -> bool {
+			<ErasStakersClipped<T>>::contains_key(&era, validator)
+		}
+
+		/// Returns the maximum number of pages of exposure we can store.
+		fn get_max_exposure_page_count() -> PageIndex {
+			return <MaxExposurePageCount<T>>::get().unwrap_or_default().max(1)
+		}
+
+		/// Returns validator commission for this era and page.
+		pub(crate) fn get_validator_commission(
+			era: EraIndex,
+			validator_stash: &T::AccountId,
+		) -> Perbill {
+			<ErasValidatorPrefs<T>>::get(&era, validator_stash).commission
+		}
+
+		/// Creates an entry to track validator reward has been claimed for a given era and page.
+		pub(crate) fn set_rewards_as_claimed(
+			era: EraIndex,
+			validator: &T::AccountId,
+			page: PageIndex,
+		) {
+			let mut claimed_pages = ClaimedRewards::<T>::get(era, validator);
+			let search = claimed_pages.binary_search(&page);
+			// this should never be called if the reward has already been claimed
+			debug_assert!(search.is_err());
+
+			match search {
+				Err(index) => {
+					claimed_pages.insert(index, page);
+					ClaimedRewards::<T>::insert(era, validator, claimed_pages);
+				},
+				_ => {},
+			}
+		}
+
+		/// Store exposure for elected validators at start of an era.
+		pub(crate) fn set_validator_exposure(
+			era: EraIndex,
+			validator: &T::AccountId,
+			exposure: Exposure<T::AccountId, BalanceOf<T>>,
+		) {
+			<ErasStakers<T>>::insert(era, &validator, &exposure);
+
+			let page_size = <ExposurePageSize<T>>::get()
+				.unwrap_or_else(|| T::MaxNominatorRewardedPerValidator::get())
+				.clamp(1, T::MaxNominatorRewardedPerValidator::get());
+			let max_page_count = Self::get_max_exposure_page_count();
+
+			let nominator_count = exposure.others.len();
+			let page_count =
+				nominator_count.saturating_add(page_size as usize - 1) / page_size as usize;
+
+			// clip nominators if it exceeds the maximum page count.
+			let exposure = if page_count as PageIndex > max_page_count {
+				// sort before clipping.
+				let mut exposure_clipped = exposure;
+				let clipped_max_len = max_page_count * page_size;
+
+				exposure_clipped.others.sort_by(|a, b| b.value.cmp(&a.value));
+				exposure_clipped.others.truncate(clipped_max_len as usize);
+				exposure_clipped
+			} else {
+				exposure
+			};
+
+			let (exposure_overview, exposure_pages) = exposure.into_pages(page_size);
```

Comment:
based on the code before, I would have expected this function to return a `Result` of some sort, which, because we are doing the clipping, should always be `Ok()`. WDYT? 

Or else, we can make `into_pages` infallible, but do the clipping inside of it?

Reply:
I feel I get what you meant but not too sure. The `into_pages` breaks down an exposure into as many pages as it needs to and so it is already infallible.

The weird part probably is to assume `exposure.into_pages` will return the required number of pages. But since I pass `page_size` to it, it also not too bad an assumption. At minimum, I can add a `debug_assert` that length of Vec<pages> is same as `required_page_count`. I will see if I can make it a bit more elegant.